src/middlewared/middlewared/plugins/rsync.py

```diff
@@ -591,3 +622,86 @@ class Config:
         res = await self.middleware.call('datastore.delete', self._config.datastore, id)
         await self.middleware.call('service.restart', 'cron')
         return res
+
+    @private
+    async def commandline(self, id):
+        """
+        Helper method to generate the rsync command avoiding code duplication.
+        """
+        rsync = await self._get_instance(id)
+        line = [
+            '/usr/bin/lockf', '-s', '-t', '0', '-k', f'"{rsync["path"]}"', '/usr/local/bin/rsync'
+        ]
+        for name, flag in (
+            ('archive', '-a'),
+            ('compress', '-z'),
+            ('delayupdates', '--delay-updates'),
+            ('delete', '--delete-delay'),
+            ('preserveattr', '-X'),
+            ('preserveperm', '-p'),
+            ('recursive', '-r'),
+            ('times', '-t'),
+        ):
+            if rsync[name]:
+                line.append(flag)
+        if rsync['extra']:
+            line.append(' '.join(rsync['extra']))
+
+        # Do not use username if one is specified in host field
+        # See #5096 for more details
+        if '@' in rsync['remotehost']:
+            remote = rsync['remotehost']
+        else:
+            remote = f'"{rsync["user"]}"@{rsync["remotehost"]}'
+
+        if rsync['mode'] == 'module':
+            module_args = [f'"{rsync["path"]}"', f'{remote}::"{rsync["remotemodule"]}"']
+            if rsync['direction'] != 'push':
+                module_args.reverse()
+            line += module_args
+        else:
+            line += [
+                '-e',
+                f'"ssh -p {rsync["remoteport"]} -o BatchMode=yes -o StrictHostKeyChecking=yes"'
+            ]
+            if pipes.quote(rsync['remotepath']) == rsync['remotepath']:
+                rsync_remotepath = rsync['remotepath']
+            else:
+                rsync_remotepath = f'\\""{rsync["remotepath"]}"\\"'
+            path_args = [f'"{rsync["path"]}"', f'{remote}:{rsync_remotepath}']
+            if rsync['direction'] != 'push':
+                path_args.reverse()
+            line += path_args
+
+        if rsync['quiet']:
+            line += ['>', '/dev/null', '2>&1']
+        return ' '.join(line)
```

Comment:
You can `return " ".join([shlex.quote(arg) for arg in line])` and avoid calling `pipes.quote` (which is obsolete method) every time